File: Freestylecoding.Math.FSharp.Tests/Natural/EqualityOperators.fs
namespace Natural

open Xunit
open Freestylecoding.Math

type public EqualityOperators() =
    static member Equality( this:'T when 'T :> System.Numerics.IEqualityOperators<'T,'T,bool>, that:'T ) : bool =
        'T.op_Equality( this, that )

    static member Inequality( this:'T when 'T :> System.Numerics.IEqualityOperators<'T,'T,bool>, that:'T ) : bool =
        'T.op_Inequality( this, that )

    [<Theory>]
    [<InlineData(  0u,  0u, true)>]
    [<InlineData(  1u,  0u, false)>]
    [<InlineData(  0u,  1u, false)>]
    [<InlineData(  1u,  1u, true)>]
    member public this.EqualitySanity left right (expected:bool) =
        Assert.Equal(
            expected,
            EqualityOperators.Equality(
                Natural( [left] ),
                Natural( [right] )
            )
        )

    [<Fact>]
    member public this.EqualityLargeNaturalsTrue () =
        Assert.True(
            EqualityOperators.Equality(
                Natural( [0xFu; 0x00000101u] ),
                Natural( [0xFu; 0x00000101u] )
            )
        )

    [<Fact>]
    member public this.EqualityLargeNaturalsFalse () =
        Assert.False(
            EqualityOperators.Equality(
                Natural( [0x8u; 0x00000101u] ),
                Natural( [0xFu; 0x00000101u] )
            )
        )

    [<Theory>]
    [<InlineData(  0u,  0u, false)>]
    [<InlineData(  1u,  0u, true)>]
    [<InlineData(  0u,  1u, true)>]
    [<InlineData(  1u,  1u, false)>]
    member public this.InequalitySanity left right (expected:bool) =
        Assert.Equal(
            expected,
            EqualityOperators.Inequality(
                Natural( [left] ),
                Natural( [right] )
            )
        )

    [<Fact>]
    member public this.InequalityLargeNaturalsFalse () =
        Assert.False(
            EqualityOperators.Inequality(
                Natural( [0xFu; 0x00000101u] ),
                Natural( [0xFu; 0x00000101u] )
            )
        )

    [<Fact>]
    member public this.InequalityLargeNaturalsTrue () =
        Assert.True(
            EqualityOperators.Inequality(
                Natural( [0x8u; 0x00000101u] ),
                Natural( [0xFu; 0x00000101u] )
            )
        )

File: Freestylecoding.Math.CSharp.Tests/Natural/EqualityOperators.cs
using System.Numerics;
using Xunit;

namespace Natural.Tests;

using static Freestylecoding.Math.CSharp.Tests.Helpers;
using Natural = Freestylecoding.Math.Natural;

public class EqualityOperators {
	private static bool InterfaceEq<T>( T left, T right ) where T : IEqualityOperators<T, T, bool> =>
		left == right;

	private static bool InterfaceNe<T>( T left, T right ) where T : IEqualityOperators<T, T, bool> =>
		left != right;

	[Theory]
	[InlineData( 0u, 0u, true )]
	[InlineData( 1u, 0u, false )]
	[InlineData( 0u, 1u, false )]
	[InlineData( 1u, 1u, true )]
	public void EqualitySanity( uint left, uint right, bool expected ) =>
		Assert.Equal(
			expected,
			InterfaceEq(
				new Natural( left ),
				new Natural( right )
			)
		);

	[Fact]
	public void EqualityLargeNaturalsTrue() =>
		Assert.True(
			InterfaceEq(
				FListNatural( 0xFu, 0x00000101u ),
				FListNatural( 0xFu, 0x00000101u )
			)
		);

	[Fact]
	public void EqualityLargeNaturalsFalse() =>
		Assert.False(
			InterfaceEq(
				FListNatural( 0x8u, 0x00000101u ),
				FListNatural( 0xFu, 0x00000101u )
			)
		);

	[Theory]
	[InlineData( 0u, 0u, false )]
	[InlineData( 1u, 0u, true )]
	[InlineData( 0u, 1u, true )]
	[InlineData( 1u, 1u, false )]
	public void InequalitySanity( uint left, uint right, bool expected ) =>
		Assert.Equal(
			expected,
			InterfaceNe(
				new Natural( left ),
				new Natural( right )
			)
		);

	[Fact]
	public void InequalityLargeNaturalsFalse() =>
		Assert.False(
			InterfaceNe(
				FListNatural( 0xFu, 0x00000101u ),
				FListNatural( 0xFu, 0x00000101u )
			)
		);

	[Fact]
	public void InequalityLargeNaturalsTrue() =>
		Assert.True(
			InterfaceNe(
				FListNatural( 0x8u, 0x00000101u ),
				FListNatural( 0xFu, 0x00000101u )
			)
		);
}